src/hooks.h

```diff
@@ -0,0 +1,255 @@
+// is used by the Kaleidoscope class and in all other places
+// where hooks are called.
+//
+class PluginHookAdapter__ {
+ public:
+
+  virtual void init() {}
+
+  virtual bool eventHandlerHook(Key &mappedKey, const EventKey &eventKey) {
+    return true;
+  }
+
+  virtual void preReportHook() {}
+  virtual void postReportHook() {}
+};
+
+
+
+// The HOOK_TASK macro defines an auxiliary Hook class (HOOK) that invokes
+// a plugin hook method with a provided set of method arguments. The
+// HOOK class is meant to be passed to the PluginLoop in the PluginHookAdapter
+// to forward the call to the hook methods of plugins.
+//
+#define HOOK_TASK(HOOK, RET_VAL, HOOK_METHOD)                                  \
+__NN__                                                                         \
+__NN__   struct HOOK {                                                         \
+__NL__                                                                         \
+__NL__      typedef RET_VAL ReturnValue;                                       \
+__NL__                                                                         \
+__NL__      template<typename Plugin__, typename... Args__>                    \
+__NL__      static RET_VAL invoke(Plugin__ &plugin, Args__&&... hook_args)     \
+__NL__      {                                                                  \
+__NL__         MATCH_HOOK_TYPE(HOOK_METHOD, Plugin__)                          \
+__NL__         return plugin.HOOK_METHOD(hook_args...);                        \
+__NL__      }                                                                  \
+__NL__   };
+
+// A predicate class that decides on hook-plugin-loop continuation.
+//
+struct ContinueIfHookReturnsTrue {
+  static bool eval(bool val) {
+    return val;
+  }
+};
+
+// This is the implementation of the PluginHookAdapter__ interface.
+// It is templated to allow it to adapt itself to the list of plugins
+// that is used in the sketch. This information is known at compile time
+// but differs between sketches as part of the firmware configuration.
+//
+// The PluginHookAdapter operates on a PluginLoop__, an
+// interface class that casts Hook classes on all plugins, therby
+// forwaring the actual hook calls. Hook overrides may
+// specify ContinuationPredicate__ classes that decide
+// if further hook methods of other plugins are allowed to be called based
+// on the return value of the most recent call to a hook, e.g. a boolean value
+// (see the eventHandlerHook for an example).
+//
+template<typename PluginLoop__>
+class PluginHookAdapter : public PluginHookAdapter__ {
+ public:
+
+  HOOK_TASK(InitTask, void, init)
+  void init() final {
+    PluginLoop__::template apply<InitTask>();
+  }
+
+  HOOK_TASK(EventHandlerHookTask, bool, eventHandlerHook)
+  bool eventHandlerHook(Key &mappedKey, const EventKey &eventKey) final {
+    return PluginLoop__::template apply<EventHandlerHookTask, ContinueIfHookReturnsTrue>
```

Comment:
Confusingly, here you've capitalized `EventHandlerHookTask`, whereas the hook name is `eventHandlerHook` (no initial capital). For pattern matching, it's less potentially confusing if the exact name of the hook is here, as for the other hooks.

Reply:
The `HOOK_TASK` generates an auxiliary struct whose name is `EventHandlerHookTask`. Structs have by convention capitalized names.

Reply:
In that case, it's the other ones that are incorrect.

Reply:
What do you mean by "the other ones"?

Reply:
Look below: `preReportHookTask` & `postReportHookTask` do not have initial capitals.

Reply:
Just added a commit that fixes this.

Reply:
Thanks; that helps. If `HOOK_TASK()` is a black box (as it is to me), it's not obvious what type of objects its parameters are, so it does help to stick to the consistent naming convention — dumb pattern matching is clear, and doesn't cause questions and confusion.